### PR TITLE
V3-Resolver

### PR DESCRIPTION
**Pull request by iri031**

### CODE REVIEW

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+                userData.healthFactor
+            ) = IPool(IPoolAddressesProvider(getPoolAddressProvider()).getPool()).getUserAccountData(user);
+        }
+
+        {
+            BaseCurrency memory baseCurr;
+            IAaveOracle aaveOracle = IAaveOracle(getAaveOracle());
+            baseCurr.baseUnit = aaveOracle.BASE_CURRENCY_UNIT();
+            baseCurr.baseAddress = (aaveOracle.BASE_CURRENCY());
+            address usd = address(0x0000000000000000000000000000000000000);
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol:195]:
```suggestion
            address usd = address(0);
```

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+        bool isolationBorrowEnabled;
+        bool isPaused;
+    }
+
+    function getTokensPrices(address[] memory tokens)
+        internal
+        view
+        returns (uint256[] memory tokenPrices, uint256 ethPrice)
+    {
+        uint256[] memory tokenPrices = IAaveOracle(getAaveOracle()).getAssetsPrices(tokens);
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol:143]:
```suggestion
        tokenPrices = IAaveOracle(getAaveOracle()).getAssetsPrices(tokens);
```

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+        (, int256 EthPrice, , , ) = (AggregatorV3Interface(getChainLinkFeed()).latestRoundData());
+        ethPrice = uint256(EthPrice);
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol]:
```suggestion
        uint256[] memory _tokenPrices = IAaveOracle(getAaveOracle()).getAssetsPrices(tokens);
        (, int256 EthPrice, , , ) = (AggregatorV3Interface(getChainLinkFeed()).latestRoundData());
        ethPrice = uint256(EthPrice);
        tokenPrices = new TokenPrice[](_tokenPrices.length);
         for (uint256 i = 0; i < _tokenPrices.length; i++) {
            tokenPrices[i] = TokenPrice(_tokenPrices[i], wmul(_tokenPrices[i], uint256(ethPrice) * 10**10));
        }
```

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        uint256 supplyCap;
+        uint256 borrowCap;
+        uint256 eModeCategory;
+        uint256 debtCeiling;
+        uint256 debtCeilingDecimals;
+        uint256 liquidationFee;
+        bool isolationBorrowEnabled;
+        bool isPaused;
+    }
+
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol:144]: struct TokenPrice {
        uint256 priceInEth;
        uint256 priceInUsd;
    }

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+    function getTokensPrices(address[] memory tokens)
+        internal
+        view
+        returns (uint256[] memory tokenPrices, uint256 ethPrice)
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol]:
```suggestion
        returns (TokenPrice[] memory tokenPrices, uint256 ethPrice)
```

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    {
+        uint256[] memory tokenPrices = IAaveOracle(getAaveOracle()).getAssetsPrices(tokens);
+        (, int256 EthPrice, , , ) = (AggregatorV3Interface(getChainLinkFeed()).latestRoundData());
+        ethPrice = uint256(EthPrice);
+    }
+
+    function getEmodePrices(address[] memory tokens, address priceOracleAddr)
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol]: Update it similar to `getTokensPrices`

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+    {
+        tokenPrices = new uint256[](tokens.length);
+        for (uint256 i = 0; i < tokens.length; i++) {
+            uint256 priceBase = IPriceOracle(priceOracleAddr).getAssetPrice(tokens[i]);
+            tokenPrices[i] = priceBase;
+        }
+    }
+
+    function getPendingRewards(
+        IAaveProtocolDataProvider aaveData,
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol]: Why are we passing it in here?

Comment (pradyuman-verma) [reply contracts/protocols/mainnet/aave_v3/helpers.sol]: Use this to get the pool data provider address - https://docs.aave.com/developers/core-contracts/pooladdressesprovider#getpooldataprovider

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+            ) = IPool(IPoolAddressesProvider(getPoolAddressProvider()).getPool()).getUserAccountData(user);
+        }
+
+        {
+            BaseCurrency memory baseCurr;
+            IAaveOracle aaveOracle = IAaveOracle(getAaveOracle());
+            baseCurr.baseUnit = aaveOracle.BASE_CURRENCY_UNIT();
+            baseCurr.baseAddress = (aaveOracle.BASE_CURRENCY());
+            address usd = address(0x0000000000000000000000000000000000000);
+            if (aaveOracle.BASE_CURRENCY() == usd) {
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol]:
```suggestion
            if (aaveOracle.BASE_CURRENCY() == address(0)) {
```

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+            ) = IPool(IPoolAddressesProvider(getPoolAddressProvider()).getPool()).getUserAccountData(user);
+        }
+
+        {
+            BaseCurrency memory baseCurr;
+            IAaveOracle aaveOracle = IAaveOracle(getAaveOracle());
+            baseCurr.baseUnit = aaveOracle.BASE_CURRENCY_UNIT();
+            baseCurr.baseAddress = (aaveOracle.BASE_CURRENCY());
+            address usd = address(0);
+            if (aaveOracle.BASE_CURRENCY() == usd) {
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol]:
```suggestion
            if (aaveOracle.BASE_CURRENCY() == address(0) {
```

File: contracts/protocols/mainnet/aave_v3/helpers.sol (hunk 3 above)
Comment (iri031) [reply contracts/protocols/mainnet/aave_v3/helpers.sol]: The returned prices are in Base_currency though the units are wei only, can we apply ETH/USD conversion here?

**Comment by gabriellahayden** [comment]
Are you still having issues with your v3?

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 10]
```diff
@@ -133,26 +137,25 @@ contract AaveV3Helper is DSMath {
...
         uint256 debtCeilingDecimals;
         uint256 liquidationFee;
+        // uint256 isolationModeTotalDebt;
         bool isolationBorrowEnabled;
         bool isPaused;
     }
 
     struct TokenPrice {
-        uint256 priceInEth;
+        uint256 priceInBase;
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol]: It should be `priceInEth`

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 11]
```diff
@@ -133,26 +137,25 @@ contract AaveV3Helper is DSMath {
...
-    function getTokensPrices(address[] memory tokens)
+    function getTokensPrices(address[] memory tokens, uint256 basePriceInUSD)
         internal
         view
-        returns (TokenPrice[] memory tokenPrices, uint256 ethPrice)
+        returns (TokenPrice[] memory tokenPrices)
     {
         uint256[] memory _tokenPrices = IAaveOracle(getAaveOracle()).getAssetsPrices(tokens);
-        (, int256 EthPrice, , , ) = (AggregatorV3Interface(getChainLinkFeed()).latestRoundData());
-        ethPrice = uint256(EthPrice);
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol]: Keep it as well.

File: contracts/protocols/mainnet/aave_v3/interfaces.sol [hunk 12]
```diff
@@ -15,6 +15,23 @@ interface IPool {
         address priceSource;
         string label;
     }
+    struct ReserveData {
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/interfaces.sol]: Just move all the structs out of interfaces.

File: contracts/protocols/mainnet/aave_v3/helpers.sol [hunk 13]
```diff
@@ -133,26 +137,25 @@ contract AaveV3Helper is DSMath {
...
         tokenPrices = new TokenPrice[](_tokenPrices.length);
         for (uint256 i = 0; i < _tokenPrices.length; i++) {
-            tokenPrices[i] = TokenPrice(_tokenPrices[i], wmul(_tokenPrices[i], ethPrice * 10**10));
+            tokenPrices[i] = TokenPrice(_tokenPrices[i], wmul(_tokenPrices[i], basePriceInUSD * 10**10));
```
Comment (pradyuman-verma) [inline contracts/protocols/mainnet/aave_v3/helpers.sol]: Just try this suggestion.
```suggestion
           uint256 _tokenPriceinETH = (_tokenPrices[i] * basePriceInUSD) / ethPrice;
            tokenPrices[i] = TokenPrice(_tokenPriceinETH , wmul(_tokenPrices[i], basePriceInUSD));
```

**Pull request closed without merging**